Move Autograd to an alias dispatch key

ailzhang opened the pull request:
This PR moves `DispatchKey::Autograd` to an alias dispatch key mapping to `AutogradCPU, AutogradCUDA, AutogradXLA, AutogradOther, AutogradPrivate*` keys. 

A few things are handled in this PR:
- Update alias dispatch key mapping and precompute dispatchTable logic
- Move `Autograd` key from `always_included` set to TensorImpl constructor.
- Update `dummyTensor` constructor to take `requires_grad` as optional argument so that it's closer to the real application in op_registration_test. 
- Use `BackendSelect` key for both backend select before and after autograd layer. (1 liner in backend_select codegen)

A few planned followups ordered by priority:
- [cleanup] Update `test_dispatch.py` to include testing `Autograd`.
- [cleanup] Add Math alias key and move catchAll to Math. (to remove 2.2 in `computeDispatchTableEntryWithDebug`)
- [new feature] Add support for Math in native_functions.yaml
- [cleanup] Add iterator like functionality to DispatchKeySet
- [cleanup/large] Only add Autograd backend keys when tensor requires grad. (cc: @ljk53 ?)

ailzhang commented on c10/core/TensorImpl.cpp:
```diff
@@ -60,14 +60,26 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
+  // a backend DispatchKey and an AutogradBackend key.
+  // We automatically add the corresponding autograd key to key_set_ so that backends can stay
+  // in the old way of only registering with backend key like DispatchKey::CPU.
+  for (uint8_t i = static_cast<uint8_t>(DispatchKey::CPU); i < static_cast<uint8_t>(DispatchKey::NumBackendKeys); i++) {
```
This logic adding autograd keys is a bit hacky. Optionally we can move this to caller sites if it's a better design choice. (either in this PR or a followup)

ailzhang replied:
Ah actually the failed cpp tests are related to this logic that we now automatically added AutogradKeys in TensorImpl. But cpp tests were tested without the global `Autograd` key enable set. So where we want to put this logic will affect how we fix the cpp tests. 
Should be an easy fix either way but would like to hear more thoughts on what's better from API design perspective. :D

bhosmer replied:
I think it's better to add the autograd key here rather than make call sites do it, but it feels weird to have to do a linear pass through the backend keys, do we really have to? Seems like there's some mask-and-shift fanciness we could do here, but admittedly I haven't thought it through.

Also I'd suggest having a utility function we just call here, `getAutogradKeyForBackend(key_set)` that does the actual work. (I'm basing that on the assumption we'll be doing this elsewhere, but I'm thinking we will, right?)

ezyang replied:
Mask-and-shift would make this code branchless. Mask out everything except (e.g., CPU) bit, and then shift it to its AutogradCPU position, then OR it into the original site.

ezyang replied:
And then you could get rid of `NumBackendKeys`

bhosmer reviewed (commented):
Hi Ailing, looking good! Just a few comments - not done, I'll have to come back to finish the review later, but wanted to leave what I had so far :)

ezyang commented:
> Please only review the second commit. This first commit is submitted in #43047. (I should get a devserver and used ghstack :P)

Worth having a more elaborate PR description here

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -340,6 +340,8 @@ template<class... Args> inline void unused_arg_(const Args&...) {}
 inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(Args...)>& op, DispatchKey dispatchKey, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
+  // No alias dispatch key is allowed at runtime.
+  TORCH_INTERNAL_ASSERT(dispatchKey < DispatchKey::NumDispatchKeys);
```
`DEBUG_ONLY` plz

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -184,7 +184,9 @@ class CAFFE2_API OperatorEntry final {
-  std::array<KernelFunction, static_cast<uint8_t>(DispatchKey::NumDispatchKeys)> dispatchTable_;
+  // Alias dispatch keys have entries in dispatchTable_ and they participate in the dispatch table
+  // computation, but they're not directly called in runtime.
+  std::array<KernelFunction, static_cast<uint8_t>(DispatchKey::NumDispatchKeysWithAlias)> dispatchTable_;
```
Hmm, I find this a bit surprising. Alias keys don't participate in runtime. So shouldn't they NOT be included in the size of the dispatch table? (Since you'll never actually go to those entries.)

ailzhang replied:
Hmmm although they won't be called in runtime, I think they should still be included in the dispatch table since they participate in the computation? E.g if we register new kernels and need to update dispatch table, we'll need to look at alias keys.

ezyang replied:
But remember we redundantly store all the keys in the flat_hash_map before. And you shouldn't be looking in `dispatchTable_` when you do computation anyway.

ezyang commented on c10/core/DispatchKey.cpp:
```diff
@@ -63,30 +63,33 @@ const char* toString(DispatchKey t) {
+      return "AutogradCUDA";
+    case DispatchKey::AutogradXLA:
+      return "AutogradXLA";
+    case DispatchKey::PrivateUse1_PreAutograd:
```
Probably need to rename these too at some point.

ezyang replied:
Also, I remember you mentioning that we could have an arbitrary usage of `PrivateUse1_PreAutograd`, e.g., to intercept autograd for any dispatch key, not just `PrivateUse1`. But it may be easier to understand and simpler to require `PrivateUse1Autograd` be used for `PrivateUse1`.

ailzhang replied:
Yea was thinking about renaming these keys together with XLA but I'm not sure who'll be broken by it. Maybe I can open an issue for it to notify users and change it later?

ezyang replied:
Sure. At the very least nested tensor will stop working. cc @cpuhrsch 

You could always add aliases for the legacy names at the bottom of DispatchKey to maintain BC too.

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -107,6 +108,8 @@ enum class DispatchKey : uint8_t {
   PrivateUse2,
   PrivateUse3,
 
+  NumBackendKeys, // Sentinel, end of backend keys.
```
The enum here uses up space in the runtime dispatch table; now you will always have an (unnecessary) gap between PrivateUse3 and Meta.  This probably doesn't matter too much (since PrivateUse1 is already a bunch of extra space) but it would be nice to mention it in case we need to penny pinch space later. Note that you can't conveniently say `NumBackendKeys = PrivateUse3 + 1` as that will cause `Meta` to get offset.

ezyang commented on aten/src/ATen/templates/TypeDefault.cpp:
```diff
@@ -23,6 +23,18 @@ namespace TypeDefault {
+
+TORCH_LIBRARY_IMPL(_, AutogradOther, m) {
+  m.fallback(torch::CppFunction::makeFallthrough());
+}
```
These probably don't belong in the TypeDefault.cpp template. My guess for where they should go, is they should go in the same place the historic Autograd fallthrough lived.

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -291,6 +319,35 @@ C10_API std::ostream& operator<<(std::ostream&, DispatchKey);
+  return k > DispatchKey::NumDispatchKeys && k < DispatchKey::NumDispatchKeysWithAlias;
+}
+
+static inline std::array<DispatchKey, 7> resolveAliasAutogradKey() {
```
Yeah, `std::array<DispatchKey, 7>` type here is weird (although, it is true that it is the same size as a DispatchKeySet!)

ezyang replied:
I guess you didn't do `DispatchKeySet` here because you don't have `DispatchKeySet` defined at this point in time. It's probably not a big deal if you just put these definitions in `DispatchKeySet.h`

ailzhang replied:
Haha yea, the main reason I put it here is mainly that our DispatchKeySet is not a iterator-like object. Per Ed's comment above, it's worth adding it to DispatchKeySet now so I'll probably do that ;)

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -51,9 +51,22 @@ class DispatchKeySet final {
+        repr_ = t == DispatchKey::Undefined ? 0
+                : 1ULL << (static_cast<uint8_t>(t) - 1);
+      }
+    }
```
This would also solve the Undefined special case, since you'd just return empty set in that case

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -126,17 +139,29 @@ static inline DispatchKey legacyExtractDispatchKey(DispatchKeySet s) {
+  // is the most likely key that will need this treatment;
+  // After Autograd keys are moved from globally enabled set to TensorImpl,
+  // we should remove all Autograd keys before taking highestPriority.
+  return (s - DispatchKeySet(DispatchKey::Autograd)).highestPriorityTypeId();
```
Good times. Reminds me that yes, we should get rid of this some time haha

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -155,24 +155,49 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  //  (4) Use catchAll kernel if available
+  // XXX: currently Autograd is the only alias key, we'll update alias key precedence after we add new
+  //      alias keys AutogradDispatchCPUOrCUDA and Math.
+  // TODO: we can remove (2.2) and (4) after TypeDefault registrations are moved from catchAll to Math.
```
ok good. 2.2 special case is pretty weird

ailzhang replied:
After trying out a few things I feel this is actually the most contained change to make it work for now and it can be easily removed once we have catchAll moved to Math. 
I'm thinking about keeping 2.2 in this PR and making removing catchAll as next task, wdyt?

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -182,6 +207,12 @@ void OperatorEntry::updateDispatchTable_(const c10::Dispatcher& dispatcher, Disp
   dispatchKeyExtractor_.setOperatorHasFallthroughForKey(dispatch_key, dispatchTable_[dispatch_ix].isFallthrough());
+  // Updates to alias dispatch key should trigger recomputation of all runtime keys mapped to the alias.
+  if (at::isAliasDispatchKey(dispatch_key)) {
+    for (auto k: c10::resolveAliasAutogradKey()) {
```
Oh, I guess this is a case where having an iterator over the alias set is useful. I guess it's time to add iterator like functionality to dispatch key set...

ailzhang replied:
Yea created #43441 as a followup item for it :D

ailzhang commented:
> This looks pretty good. I'm a little interested in testing strategy; did `test_dispatch.py` catch bugs while this was in development? If we forget to update entries related to alias dispatch keys, does the test suite catch that bug?

Yep `test_dispatch.py` did catch bugs! I didn't know about this file so didn't run it locally before sending the PR :P. I updated the file to make it include testing autograd keys, please let me know if it's still missing coverage.

bhosmer commented on c10/core/DispatchKey.h:
```diff
@@ -180,40 +183,47 @@ enum class DispatchKey : uint8_t {
-  // first.  You generally should NOT redispatch after handling autograd
-  // here (since that would result in execution of the Autograd
-  // operator, which you're trying to skip).  In PreAutograd implementations,
+  // Autograd is now an alias dispatch which by default maps to all
```
typo: missing "key" after "dispatch" :P

bhosmer commented on c10/core/DispatchKey.h:
```diff
@@ -269,13 +296,13 @@ enum class DispatchKey : uint8_t {
-// operators are not appropriate.
+// PyTorch for operators you define through AutogradOther.  Autograd backend
+// DispatchKeys serve as "wrapper" DispatchKeys: they are most appropriate for
+// tensors that compose multiple internal tensors, and for cases when the
```
I'm not sure what this means...

bhosmer commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -2,16 +2,25 @@
+DispatchKeySet getRuntimeDispatchKeySet(DispatchKey t) {
+  switch (t) {
+    case DispatchKey::Autograd:
+      return DispatchKeySet({DispatchKey::AutogradCPU,
```
Probably worth precomputing this rather than building it every time?

bhosmer commented on c10/core/DispatchKeySet.h:
```diff
@@ -108,6 +116,11 @@ class DispatchKeySet final {
   }
+
+  DispatchKey highestPriorityBackendTypeId() const {
+    return (*this & DispatchKeySet(DispatchKeySet::FULL_AFTER_INCLUSIVE, DispatchKey::PrivateUse3))
```
Might be worth having an alias defined alongside the dispatch keys instead of using `PrivateUse3` directly, just to encapsulate.

bhosmer commented on c10/core/TensorOptions.h:
```diff
@@ -621,9 +621,9 @@ inline DispatchKey computeDispatchKey(TensorOptions options) {
 
 inline DeviceType computeDeviceType(DispatchKey tid) {
-  if (tid == DispatchKey::CPU) {
+  if (tid == DispatchKey::CPU || tid == DispatchKey::AutogradCPU) {
```
I'm a little worried about using `AutogradCPU` and `AutogradCUDA` and `AutogradXLA` to detect their respective backends - it adds an asymmetry where backends without backend-specific autograd keys aren't visible to the system in the same way, so if somebody builds logic that depends on detecting backend from *only* the autograd key, it's only going to work for some backends. I think it might be more consistent just to ignore Autograd keys when detecting backend. WDYT?

bhosmer reviewed (commented):
Hi Ailing - looking excellent, got a few things inline I'm curious your thoughts on, plus some more minor comments :)

ailzhang commented on aten/src/ATen/core/VariableFallbackKernel.cpp:
```diff
@@ -28,8 +28,17 @@ using c10::KernelFunction;
 namespace {
 
-TORCH_LIBRARY_IMPL(_, Autograd, m) {
+// We only register fallback for in-tree backends here and leave it to
```
Good point - I agree should just register it for all Autograd keys here.

ezyang commented on aten/src/ATen/core/boxing/impl/test_helpers.h:
```diff
@@ -24,11 +24,18 @@ inline at::Tensor dummyTensor(c10::DispatchKeySet ks) {
+  at::Tensor t = at::detail::make_tensor<c10::TensorImpl>(storage_impl, ks, dtype);
+  // TODO: We currently have this hack since by default all tensors have a Autograd backend key
+  //       added in TensorImpl constructor. We should remove this hack once we move to only have
+  //       Autograd backend key when the tensor requires_grad=true.
```
Hmm. I'm guessing you added this to make some of the tests work out, but it is kind of confusing to have the tests exercising a feature (tensors without the autograd key on them) which don't reflect how the actual system works. I wouldn't even call it a hack... more like, these tests come from the future :>

ezyang commented on c10/core/DispatchKey.cpp:
```diff
@@ -108,4 +111,37 @@ std::ostream& operator<<(std::ostream& str, DispatchKey rhs) {
+  }
+}
+
+std::vector<DispatchKey> getRuntimeDispatchKeys(DispatchKey k) {
```
there isn't any reason for this to return a vector by value. You can `static` initialize the vectors and then return an ArrayRef, since these never change.

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -108,6 +116,11 @@ class DispatchKeySet final {
     return static_cast<DispatchKey>(64 - llvm::countLeadingZeros(repr_));
   }
+
+  DispatchKey highestPriorityBackendTypeId() const {
```
Hmm, it's a bit unexpected to need this; you could unfuse the "mask with `DispatchKeySet(DispatchKeySet::FULL_AFTER_INCLUSIVE, DispatchKey::EndOfBackendKeys)`" operation. I'll keep an eye out for its usage site.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -60,14 +60,22 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
+  // in the old way of only registering with backend key like DispatchKey::CPU.
+  // TODO: Ideally this logic fits best in Variable/Autograd layer so that we only
+  // add AutogradBackend key when the tensor requires grad.
+  DispatchKey k = key_set.highestPriorityBackendTypeId();
```
Ah, I see what you're using this for. There's an implicit invariant here that there is actually only one backend type id (the choice of highest priority one is for ease of computation). I think I agree this is probably the easiest way to implement this.

ezyang commented on test/test_dispatch.py:
```diff
@@ -249,13 +255,16 @@ def test_impl_only(self):
 schema: (none)
-Autograd: impl_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
+CPU: impl_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
+AutogradCPU: impl_t_t :: (Tensor _0) -> (Tensor _0) [ boxed unboxed ]
```
It would probably be good to actually test the alias here at some point. Especially what happens if you have both "autograd" and "autogradcpu"

ezyang reviewed (approved):
Some of the testing choices seem a bit questionable, but the core logic looks good. Ship it if everything is passing.

bhosmer reviewed (approved):
🎸 🤘

ezyang commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -2,16 +2,73 @@
+// we destruct the dispatch table.
+// http://www.cs.technion.ac.il/users/yechiel/c++-faq/construct-on-first-use-v2.html
+static DispatchKeySet& autograd_dispatch_keys (void) {
+  static DispatchKeySet* autograd_keys = new DispatchKeySet({
```
I think we can just define this as constexpr and then we don't have to worry about static initialization at all. If the initializer list DispatchKeySet constructor is not constexpr, it should be made constexpr.

ezyang replied:
(The destructor order business is nasty and I didn't realize it would be a problem, sorry!)

ailzhang commented:
@ezyang @bhosmer I fixed the internal test failures and want to make sure you're aware of the additional changes I pushed. :D 
1. The mobile failure is due to we skip registration to Autograd keys in mobile and skip the assert for mobile. It's simple fix that we just need to make sure we don't dispatch to Autograd backend keys when it's not required. [`f86086f` (#43070)](https://github.com/pytorch/pytorch/pull/43070/commits/f86086f707a7522c15230d25fbacba94bd5e6c96)
2. The second fix is a bit more involved as we have a use case described in  https://github.com/pytorch/pytorch/pull/43070/files#diff-9ce4d20732d7b4c47e6a95c566f83ff6R45-R61. From my understanding, we're required to handle this in runtime instead of dispatch table precomputation. 
After fixing the above two issues all tests are passing, I'm rebasing on top of codegen change, but want to make sure you're aware of the second issue.

ezyang commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -42,6 +42,23 @@ static inline DispatchKey dispatchTypeId(
+  // one Autograd dispatch key.
+  DispatchKeySet autograd_keys = getRuntimeDispatchKeySet(c10::DispatchKey::Autograd);
+  DispatchKey highest_autograd = (ks & autograd_keys).highestPriorityTypeId();
+  ks = (ks - autograd_keys).add(highest_autograd);
```
This looks very bad. We should work hard to avoid adding all of this extra work in the runtime dispatch path.

bhosmer replied:
If we can guarantee that (using Ailing's comment example) an autograd key of `AutogradPrivateUse1` will fall through to `PrivateUse1`, then we can precompute I think.

ezyang replied:
In VC, we established that this would not work, as AutogradOther is a one to many mapping. (I also argued that if we ever added more keys between backend and autograd, this would also not work).

bhosmer replied:
Agreed, precomputation wouldn't work for `AutogradOther` (which we decided should be used instead of `AutogradPrivateUse1` in this case, for reasons outside the scope of this PR) bc it's one-to-many. 

We also decided that the desired end state is not to use/allow fallthrough kernels on autograd keys at all, so the following no longer bears directly on this PR, but just to be a little more precise about the conditions where fallthrough precomputation would work so it doesn't go down the memory hole:
* one-to-many mappings _can't_ work (since they break the guarantee mentioned in the earlier comment, that a fallthrough's destination kernel be uniquely determined). One-to-one and many-to-one mappings can.
    * you could actually generalize this to only require that the "final destination" of a fallthrough needs to be uniquely determined - precomputation could fall through any number of non-unique intermediate stages, as long as they were populated entirely by fallthroughs, and all paths terminated at the same non-fallthrough kernel.
* new intermediate stages (e.g. "more keys between backend and autograd") that preserve the first condition _can_ work, but would need an upgrade to the precomputation algorithm to teach it the new keys.

ailzhang replied:
Btw created https://github.com/pytorch/pytorch/issues/43908 to track that issue. IIRC you mentioned there're some boxed/unboxed stuff blocking us moving to the redispatch as fallback Autograd kernel? Would you mind updating that thread with some context? ;) Thanks a lot!

Commit:
Move Autograd to an alias dispatch key (#43070)

Summary:
This PR moves `DispatchKey::Autograd` to an alias dispatch key mapping to `AutogradCPU, AutogradCUDA, AutogradXLA, AutogradOther, AutogradPrivate*` keys.

A few things are handled in this PR:
- Update alias dispatch key mapping and precompute dispatchTable logic
- Move `Autograd` key from `always_included` set to TensorImpl constructor.
- Update `dummyTensor` constructor to take `requires_grad` as optional argument so that it's closer to the real application in op_registration_test.
- Use `BackendSelect` key for both backend select before and after autograd layer. (1 liner in backend_select codegen)

A few planned followups ordered by priority:
- [cleanup] Update `test_dispatch.py` to include testing `Autograd`.
- [cleanup] Add Math alias key and move catchAll to Math. (to remove 2.2 in `computeDispatchTableEntryWithDebug`)
- [new feature] Add support for Math in native_functions.yaml
- [cleanup] Add iterator like functionality to DispatchKeySet
- [cleanup/large] Only add Autograd backend keys when tensor requires grad. (cc: ljk53 ?)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/43070

Reviewed By: ezyang

Differential Revision: D23281535

Pulled By: ailzhang

fbshipit-source-id: bdda523ccfcfd13977bc40e98b00f4c6cd6cd980
